test/Npgsql.Tests/CommandTests.cs

```diff
@@ -1052,6 +1050,35 @@ public async Task SameCommandDifferentParamInstances()
             Assert.That(await cmd.ExecuteScalarAsync(), Is.EqualTo(9));
         }
 
+        [Test, IssueLink("https://github.com/npgsql/npgsql/issues/3509"), Timeout(5000)]
+        public async Task Bug3509()
+        {
+            if (IsMultiplexing)
+                return;
+
+            var csb = new NpgsqlConnectionStringBuilder(ConnectionString)
+            {
+                KeepAlive = 1,
+            };
+            await using var postmasterMock = PgPostmasterMock.Start(csb.ToString());
+            using var _ = CreateTempPool(postmasterMock.ConnectionString, out var connectionString);
+            await using var conn = await OpenConnectionAsync(connectionString);
+            var serverMock = await postmasterMock.WaitForServerConnection();
+            await serverMock.WaitForQuery();
+            var queryTask = Task.Run(async () => await conn.ExecuteNonQueryAsync("SELECT 1"));
+            // giving a queryTask some time to get stuck on a lock
+            await Task.Delay(100);
```

Comment:
This kind of thing creates flaky tests, especially in CI... There's just no good way to reliably test concurrency issues.

But we can try and see.